pkg/controller/infrastructure/actuator_reconcile.go

```diff
@@ -391,6 +411,112 @@ func createFirewall(ctx context.Context, r *firewallReconciler) error {
 	return updateProviderStatus(ctx, r.c, r.infrastructure, r.providerStatus, &nodeCIDR)
 }
 
+func reconcileEgressIPs(ctx context.Context, r *egressIPReconciler) error {
+	static := metalgo.IPTypeStatic
+	currentEgressIPs := sets.NewString()
+	resp, err := r.mclient.IPFind(&metalgo.IPFindRequest{
+		ProjectID: &r.infrastructureConfig.ProjectID,
+		Tags:      []string{r.egressTag},
+		Type:      &static,
+	})
+	if err != nil {
+		return &controllererrors.RequeueAfterError{
+			Cause:        errors.Wrap(err, "failed to list egress ips of cluster"),
+			RequeueAfter: 30 * time.Second,
+		}
+	}
+	for _, ip := range resp.IPs {
+		currentEgressIPs.Insert(*ip.Ipaddress)
+	}
+
+	wantEgressIPs := sets.NewString()
+	for _, egressRule := range r.infrastructureConfig.Firewall.EgressRules {
+		wantEgressIPs.Insert(egressRule.IPs...)
+
+		for _, ip := range egressRule.IPs {
+			if currentEgressIPs.Has(ip) {
+				continue
+			}
+
+			resp, err := r.mclient.IPFind(&metalgo.IPFindRequest{
+				IPAddress: &ip,
+				ProjectID: &r.infrastructureConfig.ProjectID,
+				NetworkID: &egressRule.NetworkID,
+			})
+
+			if err != nil {
+				return &controllererrors.RequeueAfterError{
+					Cause:        errors.Wrap(err, fmt.Sprintf("error when retrieving ip %s for egress rule", ip)),
+					RequeueAfter: 30 * time.Second,
+				}
+			}
+
+			if len(resp.IPs) == 0 || len(resp.IPs) > 1 {
+				return &controllererrors.RequeueAfterError{
+					Cause:        errors.Wrap(err, fmt.Sprintf("ip %s for egress rule is ambiguous", ip)),
+					RequeueAfter: 30 * time.Second,
+				}
+			}
+
+			dbIP := resp.IPs[0]
+			if dbIP.Type != nil && *dbIP.Type != metalgo.IPTypeStatic {
+				return &controllererrors.RequeueAfterError{
+					Cause:        errors.Wrap(err, fmt.Sprintf("ips for egress rule must be static, but %s is not static", ip)),
+					RequeueAfter: 30 * time.Second,
+				}
+			}
+
+			if len(dbIP.Tags) > 0 {
```

Comment:
From the error message it says that the IP does not have an egress tag. Where is this check if this IP has this kind of tag or not? And if it already has one, couldn't the length of `dbIP.Tags` be `1` as well?

Reply:
Ah, I think I got it, it's the first condition where you `continue`. So, we will somewhere need to document that a user is not allowed to use tags on IPs when he wants to use the SNAT feature. Could be nice to allow custom user tags in the future, I guess.

Reply:
If the IP has just a single egress tag for this cluster, it's already contained in `currentEgressIPs`: https://github.com/metal-stack/gardener-extension-provider-metal/blob/f62fb633a38d8075eee4c20168962be5e4eb6af9/pkg/controller/infrastructure/actuator_reconcile.go#L437
If the IP has some tags at the given line, it's assumed to be already taken for some other purpose.